emu_sv/time_evolution.py

```diff
@@ -1,32 +1,274 @@
 import torch
-
+from typing import Any, no_type_check
 from emu_base.math.krylov_exp import krylov_exp
+from emu_base.math.double_krylov import double_krylov
 from emu_sv.hamiltonian import RydbergHamiltonian
 
 
-def do_time_step(
-    dt: float,
-    omegas: torch.Tensor,
-    deltas: torch.Tensor,
-    phis: torch.Tensor,
-    full_interaction_matrix: torch.Tensor,
-    state_vector: torch.Tensor,
-    krylov_tolerance: float,
-) -> tuple[torch.Tensor, RydbergHamiltonian]:
-    ham = RydbergHamiltonian(
-        omegas=omegas,
-        deltas=deltas,
-        phis=phis,
-        interaction_matrix=full_interaction_matrix,
-        device=state_vector.device,
-    )
-    op = lambda x: -1j * dt * (ham * x)
-    return (
-        krylov_exp(
+class DHDOmegaSparse:
+    """
+    Derivative of the RydbergHamiltonian respect to Omega.
+        ∂H/∂Ωₖ = 0.5[cos(ϕₖ)σˣₖ + sin(ϕₖ)σʸₖ]
+
+    If ϕₖ=0, simplifies to ∂H/∂Ωₖ = 0.5σˣₖ
+    """
+
+    def __init__(self, index: int, device: str, nqubits: int, phi: torch.Tensor):
+        self.index = index
+        self.shape = (2**index, 2, 2 ** (nqubits - index - 1))
+        self.inds = torch.tensor([1, 0], device=device)
+        self.alpha = 0.5 * torch.exp(1j * phi).item()
+        if phi != 0:
```

Comment:
we just discovered last week that this pattern is toxic. Let's refactor

Reply:
I think the function assignment is good now cuz you are using static/global functions

I think better use something like `torch.is_nonzero` rather than `!= 0`